py/dml/traits.py

```diff
@@ -653,15 +653,14 @@ def scope(self, global_scope):
         s = Symtab(global_scope)
         selfref = mkLit(self.site, '_' + cident(self.name), self.type())
         for name in self.members():
-            # HACK the dumbness of this speaks for itself.
-            # One way to address this hack is to add a symbol class with a
-            # deferred-calculated expression.
+            # This is very hacky, but works well
             try:
-                s.add(ExpressionSymbol(
-                    name, mkSubRef(self.site, selfref, name, '.'), self.site))
+                symbol = ExpressionSymbol(
+                    name, mkSubRef(self.site, selfref, name, '.'), self.site)
             except EINDEPENDENTVIOL:
-                s.add(InvalidExpressionSymbol(name, EINDEPENDENTVIOL,
-                                              self.site))
+                symbol = InvalidExpressionSymbol(name, EINDEPENDENTVIOL,
```

Comment:
Might be more conservative to use an ExpressionSymbol with a non-value, because there's a precedent for that: That `Symbol.expr()` may throw an exception is something new, while we already do protect the invariant that standalone non-values will give errors.